central/complianceoperator/v2/rules/datastore/datastore_impl_test.go

```diff
@@ -1 +1,368 @@
+				getTestRule(testconsts.Cluster1),
+				getTestRule(testconsts.Cluster2),
+				getTestRule(testconsts.Cluster3),
+			},
+			testContext:         s.testContexts[testutils.UnrestrictedReadWriteCtx],
+			expectedRecordIndex: set.NewFrozenIntSet(0, 1, 2),
+		},
+		{
+			desc: "Write 3 clusters - No access",
+			rules: []*storage.ComplianceOperatorRuleV2{
+				getTestRule(testconsts.Cluster1),
+				getTestRule(testconsts.Cluster2),
+				getTestRule(testconsts.Cluster3),
+			},
+			testContext:         s.noAccessCtx,
+			expectedRecordIndex: set.NewFrozenIntSet(),
+		},
+		{
+			desc: "Write 3 clusters - Cluster 1 access",
+			rules: []*storage.ComplianceOperatorRuleV2{
+				getTestRule(testconsts.Cluster1),
+				getTestRule(testconsts.Cluster2),
+				getTestRule(testconsts.Cluster3),
+			},
+			testContext:         s.testContexts[testutils.Cluster1ReadWriteCtx],
+			expectedRecordIndex: set.NewFrozenIntSet(0),
+		},
+	}
+
+	for _, tc := range testCases {
+		for index, rule := range tc.rules {
+			if tc.expectedRecordIndex.Contains(index) {
+				s.Require().NoError(s.dataStore.UpsertRule(tc.testContext, rule))
+			} else {
+				s.Require().Error(s.dataStore.UpsertRule(tc.testContext, rule), "access to resource denied")
+			}
+		}
+
+		count, err := s.storage.Count(s.hasReadCtx)
+		s.Require().NoError(err)
+		s.Require().Equal(tc.expectedRecordIndex.Cardinality(), count)
+
+		// Clean up
+		for _, rule := range tc.rules {
+			s.Require().NoError(s.dataStore.DeleteRule(s.hasWriteCtx, rule.GetId()))
+		}
+	}
+}
+
+func (s *complianceRuleDataStoreTestSuite) TestUpsertRules() {
+	// make sure we have nothing
+	ruleIDs, err := s.storage.GetIDs(s.hasReadCtx)
+	s.Require().NoError(err)
+	s.Require().Empty(ruleIDs)
+
+	testCases := []struct {
+		desc                string
+		rules               []*storage.ComplianceOperatorRuleV2
+		testContext         context.Context
+		expectedRecordIndex set.FrozenIntSet
+	}{
+		{
+			desc: "Write 3 clusters - Full access",
+			rules: []*storage.ComplianceOperatorRuleV2{
+				getTestRule(testconsts.Cluster1),
+				getTestRule(testconsts.Cluster2),
+				getTestRule(testconsts.Cluster3),
+			},
+			testContext:         s.testContexts[testutils.UnrestrictedReadWriteCtx],
+			expectedRecordIndex: set.NewFrozenIntSet(0, 1, 2),
+		},
+		{
+			desc: "Write 3 clusters - No access",
+			rules: []*storage.ComplianceOperatorRuleV2{
+				getTestRule(testconsts.Cluster1),
+				getTestRule(testconsts.Cluster2),
+				getTestRule(testconsts.Cluster3),
+			},
+			testContext:         s.noAccessCtx,
+			expectedRecordIndex: set.NewFrozenIntSet(),
+		},
+		{
+			desc: "Write 3 clusters - Cluster 1 access",
+			rules: []*storage.ComplianceOperatorRuleV2{
+				getTestRule(testconsts.Cluster1),
+				getTestRule(testconsts.Cluster2),
+				getTestRule(testconsts.Cluster3),
+			},
+			testContext:         s.testContexts[testutils.Cluster1ReadWriteCtx],
+			expectedRecordIndex: set.NewFrozenIntSet(),
```

Comment:
How come we assert this is an empty set when cluster 1 has r/w access?

Reply:
This is the case of writing multiple rules in 1 call.  We don't want a partial success.  So either they all succeed or we fail the whole thing.  Since we only have access to cluster 1, we cannot write 2 and 3 and thus fail the whole thing.